homeassistant/components/insteon_plm.py

```diff
@@ -26,95 +26,142 @@
 CONFIG_SCHEMA = vol.Schema({
     DOMAIN: vol.Schema({
         vol.Required(CONF_PORT): cv.string,
-        vol.Optional(CONF_OVERRIDE, default=[]): cv.ensure_list_csv,
-    })
+        vol.Optional(CONF_OVERRIDE): vol.All(
+            cv.ensure_list_csv, vol.Length(min=1))
+        })
 }, extra=vol.ALLOW_EXTRA)
 
-PLM_PLATFORMS = {
-    'binary_sensor': ['binary_sensor'],
-    'light': ['light'],
-    'switch': ['switch'],
-}
-
 
 @asyncio.coroutine
 def async_setup(hass, config):
     """Set up the connection to the PLM."""
     import insteonplm
 
+    ipdb = IPDB()
+
     conf = config[DOMAIN]
     port = conf.get(CONF_PORT)
-    overrides = conf.get(CONF_OVERRIDE)
+    overrides = conf.get(CONF_OVERRIDE, [])
 
     @callback
     def async_plm_new_device(device):
         """Detect device from transport to be delegated to platform."""
-        name = device.get('address')
-        address = device.get('address_hex')
-        capabilities = device.get('capabilities', [])
-
-        _LOGGER.info("New INSTEON PLM device: %s (%s) %r",
-                     name, address, capabilities)
-
-        loadlist = []
-        for platform in PLM_PLATFORMS:
-            caplist = PLM_PLATFORMS.get(platform)
-            for key in capabilities:
-                if key in caplist:
-                    loadlist.append(platform)
-
-        loadlist = sorted(set(loadlist))
-
-        for loadplatform in loadlist:
-            hass.async_add_job(
-                discovery.async_load_platform(
-                    hass, loadplatform, DOMAIN, discovered=[device],
-                    hass_config=config))
+        for state_key in device.states:
+            platform_info = ipdb[device.states[state_key]]
+            platform = platform_info.platform
+            if platform is not None:
+                _LOGGER.info("New INSTEON PLM device: %s (%s) %s",
+                             device.address,
+                             device.states[state_key].name,
+                             platform)
+
+                hass.async_add_job(
+                    discovery.async_load_platform(
+                        hass, platform, DOMAIN,
+                        discovered=[{'address': device.address.hex,
```

Comment:
Discovery info is usually a dictionary. I'm not sure if it's required to be that. Why should it be a list?

Reply:
This is a very interesting question. I left it this way because that is the way it was in the prior version of this code but looking at it, there does not seem to be a good reason for it. However, it is logical given that the callback method of async_setup_platform takes a list so it is logical to assume that discovered is a list. But you are right, it does not have to be since I am only calling one entity at a time to add to the platform. I will test it out and see if anything breaks :)